crates/resolver/src/name_server/name_server.rs

```diff
@@ -19,59 +20,53 @@ use proto::multicast::MDNS_IPV4;
 use proto::xfer::{DnsHandle, DnsRequest, DnsResponse, FirstAnswer};
 use tracing::debug;
 
-#[cfg(feature = "mdns")]
-use crate::config::Protocol;
 use crate::config::{NameServerConfig, ResolverOpts};
 use crate::error::ResolveError;
-use crate::name_server::{ConnectionProvider, NameServerState, NameServerStats};
-#[cfg(feature = "tokio-runtime")]
-use crate::name_server::{TokioConnection, TokioConnectionProvider, TokioHandle};
+use crate::name_server::{GenericConnection, NameServerState, NameServerStats, RuntimeProvider};
+#[cfg(feature = "mdns")]
+use proto::multicast::{MdnsClientConnect, MdnsClientStream, MdnsQueryType};
 
-/// Specifies the details of a remote NameServer used for lookups
+/// This struct is needed only for testing. Specifically, `C` is needed for mocking.
 #[derive(Clone)]
-pub struct NameServer<
-    C: DnsHandle<Error = ResolveError> + Send,
-    P: ConnectionProvider<Conn = C> + Send,
+pub struct AbstractNameServer<
+    C: DnsHandle<Error = ResolveError> + Send + Sync + 'static + CreateConnection,
+    P: RuntimeProvider + Send,
 > {
     config: NameServerConfig,
     options: ResolverOpts,
     client: Arc<Mutex<Option<C>>>,
     state: Arc<NameServerState>,
     stats: Arc<NameServerStats>,
-    conn_provider: P,
+    runtime_provider: P,
 }
 
-impl<C: DnsHandle<Error = ResolveError>, P: ConnectionProvider<Conn = C>> Debug
-    for NameServer<C, P>
+/// Specifies the details of a remote NameServer used for lookups
+pub type NameServer<P> = AbstractNameServer<GenericConnection, P>;
+
+impl<C, P> Debug for AbstractNameServer<C, P>
```

Comment:
I'm not sure this makes sense. Why do we have an `AbstractNameServer` and a `NameServer`? If anything, I feel like the naming here should be reversed, more like `NameServer` and `GenericNameServer` (or something like that).

Reply:
Since the original idea of `NameServer` and `GenericNameServer` comes from the need of mock testing. We expose it since the tests need it, not because we expect end users to actually use it. Rename it could make it clear in documentations.

Reply:
However, if you insist the original naming is better, I could revert it back.